inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp

```diff
@@ -1089,23 +1115,25 @@ class ScaleFactorPerLayer<InferenceEngine::WeightableLayer*> {
-            gnawarn() << "Output scale for " << wl->name
-                << " too large and are being reduced. Else saturations likely will happen \n";
-            // reduce weight scale according experimental heuristic
-            if (quant->_dst_quant.GetScale() * quant->_src_quant.GetScale() /
-                static_cast<float>(std::numeric_limits<int32_t>::max()) < _scale_change_threshold_100) {
-                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * _scale_reduction_50);
-            } else if (quant->_dst_quant.GetScale() * quant->_src_quant.GetScale() /
-                static_cast<float>(std::numeric_limits<int32_t>::max()) < _scale_change_threshold_150) {
-                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * _scale_reduction_45);
-            } else if (quant->_dst_quant.GetScale() * quant->_src_quant.GetScale() /
-                static_cast<float>(std::numeric_limits<int32_t>::max()) < _scale_change_threshold_200) {
-                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * _scale_reduction_40);
-            } else {
-                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * _scale_reduction_35);
+        if (weightsSize == 1) {
+            std::vector<std::tuple<uint16_t const, float const, float const>>::const_iterator itt = thresholds.begin();
+
+            if (inputsSize == 2 && static_cast<uint64_t>(tmp_dst_quant_scale * quant->_src_quant.GetScale()) >
+                static_cast<uint64_t>(std::numeric_limits<int32_t>::max() - 1) * std::get<0>(*itt)) {
+                gnawarn() << "Output scale for " << wl->name
+                    << " too large and are being reduced. Else saturations likely will happen \n";
+                // reduce weight scale according experimental heuristic
+                while ((itt + 1) != thresholds.end() && quant->_dst_quant.GetScale() * quant->_src_quant.GetScale() /
+                    static_cast<float>(std::numeric_limits<int32_t>::max()) >= std::get<0>(*(++itt))) {}
+                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * std::get<1>(*itt));
+            } else if (inputsSize == 1 && static_cast<uint16_t>(tmp_dst_quant_scale * quant->_src_quant.GetScale()) >
+                static_cast<uint16_t>(std::numeric_limits<int8_t>::max() - 1) * std::get<0>(*itt)) {
+                gnawarn() << "Output scale for " << wl->name
+                    << " too large and are being reduced. Else saturations likely will happen \n";
+                // reduce weight scale according experimental heuristic
+                while ((itt + 1) != thresholds.end() && quant->_dst_quant.GetScale() * quant->_src_quant.GetScale() /
+                    static_cast<float>(std::numeric_limits<int8_t>::max()) >= std::get<0>(*(++itt))) {}
+                quant->_weights_quant.SetScale(quant->_weights_quant.GetScale() * std::get<2>(*itt));
             }
```

Comment:
looks like it could be easily refactored to remove similiar code from if() and else()

Reply:
Done.